File: backend/growattAPI/src/main/java/entity/Notification.java
package entity;

import java.time.Instant;
import java.util.UUID;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

/**
 * In-app notification row in the Supabase Postgres {@code notifications} table.
 *
 * <p>Keyed by the Supabase auth user id ({@code auth_id}). Inserting a row is all the solar
 * job needs to do: a database webhook on insert invokes the {@code send-push} Edge Function,
 * which delivers the Expo push — so push is no longer sent from this service.</p>
 *
 * <p>"Mark as read" is a hard delete on the client side, so every stored row counts as unread.</p>
 */
@Entity
@Table(name = "notifications")
@Getter
@Setter
@NoArgsConstructor
public class Notification {

	@Id
	private UUID id;

	/** Owner of the notification (Supabase auth.users id). */
	@Column(name = "auth_id")
	private UUID authId;

	/** Source: 'weather_sync', 'solar_sync' or 'system'. */
	private String type;

	/** Severity: 'success', 'error', 'info' or 'warning'. */
	private String level;

	private String title;

	private String message;

	@Column(name = "created_at")
	private Instant createdAt;

	public Notification(UUID authId, String type, String level, String title, String message) {
		this.id = UUID.randomUUID();
		this.authId = authId;
		this.type = type;
		this.level = level;
		this.title = title;
		this.message = message;
		this.createdAt = Instant.now();
	}
}

File: backend/weatherAPI/models/Notification.js
const mongoose = require("mongoose");

/**
 * In-app notification (web notification center).
 *
 * Written by BOTH the Node weather backfill cron (via Mongoose) and the Java solar
 * backfill job (via Spring Data) into the SAME shared `notifications` collection, so the
 * field names and types here must stay aligned with the Java `Notification` entity.
 *
 * "Mark as read" is implemented as a hard delete, so every stored notification is, by
 * definition, unread. The unread badge count is simply the number of documents for a user.
 */
const notificationSchema = new mongoose.Schema(
  {
    // Stored as an ObjectId so it matches the Mongoose-cast `req.user` in queries and the
    // org.bson.types.ObjectId the Java job writes.
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true,
      index: true,
    },
    // Source of the notification.
    type: {
      type: String,
      enum: ["weather_sync", "solar_sync", "system"],
      default: "system",
    },
    // Outcome severity, drives the icon/colour in the UI.
    level: {
      type: String,
      enum: ["success", "error", "info", "warning"],
      default: "info",
    },
    title: { type: String, required: true },
    message: { type: String, default: "" },
    // Optional structured payload (counts, date, etc.).
    meta: { type: mongoose.Schema.Types.Mixed },
  },
  { timestamps: true },
);

// Fast "newest first, per user" look-ups.
notificationSchema.index({ userId: 1, createdAt: -1 });

const Notification = mongoose.model("Notification", notificationSchema, "notifications");

module.exports = Notification;
